tests/config_test.cc

```diff
@@ -0,0 +1,160 @@
+#include "cris/core/config/config.h"
+
+#include <gtest/gtest.h>
+
+#include <fmt/core.h>
+
+#include <chrono>
+#include <cstdint>
+#include <cstdio>
+#include <filesystem>
+#include <fstream>
+#include <thread>
+
+namespace cris::core {
+
+constexpr double kEps = 1e-6;
+
+namespace fs = std::filesystem;
+
+class ConfigTest : public testing::Test {
+   public:
+    ConfigTest()
+        : testing::Test()
+        , test_config_dir_(std::tmpnam(nullptr))
+        , test_config_path_(test_config_dir_ / "test.config") {
+        EXPECT_TRUE(fs::create_directory(test_config_dir_));
+    }
+
+    ~ConfigTest() { fs::remove_all(test_config_dir_); }
+
+    ConfigFile MakeConfigFile(std::string content) const {
+        std::ofstream config_file(test_config_path_);
+        config_file << std::move(content);
+        config_file.flush();
+        return ConfigFile(test_config_path_);
+    }
+
+   private:
+    fs::path test_config_dir_;
+    fs::path test_config_path_;
+};
+
+TEST_F(ConfigTest, Configs) {
+    const std::string int_config_name     = "int_config";
+    int               int_config_value    = 123;
+    const std::string double_config_name  = "double_config";
+    double            double_config_value = 3.14;
+
+    const std::string int_config_with_default_1_name  = "int_config_wd_1";
+    int               int_config_with_default_1_value = 23333;
+    const std::string int_config_with_default_2_name  = "int_config_wd_2";
+    int               int_config_with_default_2_value = 77777;
+
+    auto config_file = MakeConfigFile(fmt::format(
+        R"({{
+            "{}": {},
+            "{}": {},
+            "{}": {}
+        }})",
+        int_config_name,
+        int_config_value,
+        double_config_name,
+        double_config_value,
+        // wd_1 uses the default value, wd_2 uses the value in the config
+        int_config_with_default_2_name,
+        int_config_with_default_2_value));
+
+    EXPECT_EQ(config_file.Get<int>(int_config_name)->GetValue(), int_config_value);
+    EXPECT_NEAR(config_file.Get<double>(double_config_name)->GetValue(), double_config_value, kEps);
+
+    // No value in the config file, so the value passed to Register will be picked up.
+    EXPECT_EQ(
+        config_file.Get<int>(int_config_with_default_1_name, int(int_config_with_default_1_value))->GetValue(),
```

Comment:
Should we test casting within Get, e.g. calling it with the wrong type?
Including both sides, including a crashed narrowing.

Reply:
Do you mean Get<int>(k, 1.23)? No, it won't compile, because we banned implicit conversion that may lose precision.

Reply:
At least you can do the widening side.

Reply:
What do we mean?

Reply:
I mean `Get<double>(int_key, 1.23)` and `Get<int>(fp_key, 1)`.

Reply:
BTW why do you need to specify the type explicitly when it can be deducted from arg?

Reply:
- default value is not required
- Even if we have the default value, it may not be deducted to the desired type and also error-prone.